AK/OwnPtrWithCustomDeleter.h

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2022, Lucas Chollet <lucas.chollet@free.fr>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Function.h>
+#include <AK/Noncopyable.h>
+#include <AK/StdLibExtras.h>
+
+template<Pointer T>
+struct OwnPtrWithCustomDeleter {
+    AK_MAKE_NONCOPYABLE(OwnPtrWithCustomDeleter);
+
+public:
+    OwnPtrWithCustomDeleter(T ptr, Function<void(T)> deleter)
+        : m_ptr(ptr)
+        , m_deleter(move(deleter))
+    {
+    }
+
+    OwnPtrWithCustomDeleter(OwnPtrWithCustomDeleter&& other)
+    {
+        swap(m_ptr, other.m_ptr);
+        swap(m_deleter, other.m_deleter);
+    };
+
+    ~OwnPtrWithCustomDeleter()
+    {
+        if (m_ptr) {
+            VERIFY(m_deleter);
+            m_deleter(m_ptr);
```

Comment:
There's the possibility that the deleter doesn't actually free the memory, which I don't love, but I'm not sure what to do about it :thinking:

Reply:
Adding `VERIFY(m_ptr)` wouldn't actually work because the deleter freeing the memory wouldn't set `m_ptr` to null.

I had a thought when I wrote the comment that we could change the deleter callback to pass a reference to `m_ptr` instead of just the pointer itself: `Function<void(T*&)>`. Then if the deleter nulls the pointer, that actually propagates back to `m_ptr`, and we could then `VERIFY(!m_ptr)`. But that's not foolproof either, because nothing stops a deleter from just doing:

```c++
[](auto*& ptr) { ptr = nullptr; }
```

Without actually freeing the memory, so that doesn't help either.

So I guess my comment is more of a "sad memory noise" comment, and we'll have to be careful in reviews :man_shrugging:

Reply:
Passing a ref would at least warn you if you "forget", I'm going to add that.
I would consider that this lambda is intentionally harmful.

Reply:
Hmm `freeaddrinfo()` does not take a reference, so it leaves the pointer at its old state.
That means that we would need to use this kind of lambda:
```cpp
[](struct addrinfo*& ptr) {
	if (ptr) {
		::freeaddrinfo(ptr);
		ptr = nullptr;
	} 
}
```

Don't feel like it's super awesome.

Reply:
Right, just about every callback would have to do that (even if it only called `free`), so, agreed that it's extra work that doesn't actually solve the problem.